pkg/kubelet/pod_workers.go

```diff
@@ -981,26 +981,30 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
 	// enforce the restriction that a grace period can only decrease and track whatever our value is,
 	// then ensure a calculated value is passed down to lower levels
 	gracePeriod := status.gracePeriod
-	// this value is bedrock truth - the apiserver owns telling us this value calculated by apiserver
-	if override := pod.DeletionGracePeriodSeconds; override != nil {
-		if gracePeriod == 0 || *override < gracePeriod {
+
+	// defaults the grace period value to the pod's specified termination grace period seconds if present
+	// the kubelet provided no requested value (graceful termination?)
+	if override := pod.Spec.TerminationGracePeriodSeconds; override != nil {
+		if *override < gracePeriod {
 			gracePeriod = *override
 		}
 	}
+
+	// this value is top priority - the apiserver owns telling us this value calculated by apiserver
+	if override := pod.DeletionGracePeriodSeconds; override != nil {
+		gracePeriod = *override
+	}
+
 	// we allow other parts of the kubelet (namely eviction) to request this pod be terminated faster
 	if options != nil {
 		if override := options.PodTerminationGracePeriodSecondsOverride; override != nil {
-			if gracePeriod == 0 || *override < gracePeriod {
+			if *override < gracePeriod {
 				gracePeriod = *override
 			}
 		}
 	}
-	// make a best effort to default this value to the pod's desired intent, in the event
-	// the kubelet provided no requested value (graceful termination?)
-	if gracePeriod == 0 && pod.Spec.TerminationGracePeriodSeconds != nil {
```

Comment:
This was done in version 1.19. After version 1.23, it is not allowed to send a zero grace period to the CRI.

Reply:
@Seaiii Thank you for teaching me. Could you please tell me why "sending a zero grace period to the CRI is not allowed"? I'm happy if you would paste URL of discussion.

Reply:
> @Seaiii Thank you for teaching me. Could you please tell me why "sending a zero grace period to the CRI is not allowed"? I'm happy if you would paste URL of discussion.

i'm sorry. I can't find the exact article. The only thing I remember is that it was in the k8s deletion reduction time section that it said CRI doesn't want to transmit zero values. In fact, changing it to zero here won't do anything, because the second request sent to CRI will be blocked

Reply:
Thanks, it's OK to wait for somebody who know the discussion well. 
BTW, even if I fix here, return value of calculateEffectiveGracePeriod() will never be 0, always larger than 1. please see L.1009.

Reply:
> Thanks, it's OK to wait for somebody who know the discussion well. BTW, even if I fix here, return value of calculateEffectiveGracePeriod() will never be 0, always larger than 1. please see L.1009.

Yes, even if you set 0 here, it will be changed to 1 later on
You may want to take a look at this, I've asked this question before https://github.com/kubernetes/kubernetes/issues/113717